### PR TITLE
Select the dnn convolution algorithm using actually available memory

### PR DESCRIPTION
**Pull request by nouiz**
New PR rebased of gh-5305 to try to get travis passing.

### CODE REVIEW

File: theano/gpuarray/__init__.py [hunk 1]
```diff
@@ -47,65 +47,68 @@ def init_dev(dev, name=None):
...
+            init_dev.dnn_version = dnn.version()
+            # 5200 should not print warning with cudnn 5.1 final.
+            if init_dev.dnn_version >= 5200:
+                warnings.warn("Your cuDNN version is more recent than "
+                              "Theano. If you encounter problems, try "
+                              "updating Theano or downgrading cuDNN to "
+                              "version 5.1.")
+            if config.print_active_device:
+                print("Using cuDNN version %d" % init_dev.dnn_version,
+                      file=sys.stderr)
```
Comment (lamblin) [inline theano/gpuarray/__init__.py]: I'm trying to test what happens if some devices support cuDNN and some don't.

**Comment by lamblin** [comment]
So if I'm trying to run on a GTX580 (on a machine that also has a k6000), I have the following error:
```
lamblinp@eos21 % THEANO_FLAGS=device=cuda1 python -c 'import theano'
Using cuDNN version 5103
ERROR (theano.gpuarray): Could not initialize pygpu, support disabled
Traceback (most recent call last):
  File "/u/lamblinp/code/Theano/theano/gpuarray/__init__.py", line 117, in <module>
    init_dev(config.device)
  File "/u/lamblinp/code/Theano/theano/gpuarray/__init__.py", line 75, in init_dev
    context.cudnn_handle = dnn._make_handle(context)
  File "/u/lamblinp/code/Theano/theano/gpuarray/dnn.py", line 85, in _make_handle
    raise RuntimeError("error creating cudnn handle")
RuntimeError: error creating cudnn handle
```

And when trying to disable cuDNN explicitly:
```
lamblinp@eos21 % THEANO_FLAGS=device=cuda1,dnn.enabled=False python -c 'import theano'
ERROR (theano.gpuarray): Could not initialize pygpu, support disabled
Traceback (most recent call last):
  File "/u/lamblinp/code/Theano/theano/gpuarray/__init__.py", line 117, in <module>
    init_dev(config.device)
  File "/u/lamblinp/code/Theano/theano/gpuarray/__init__.py", line 75, in init_dev
    context.cudnn_handle = dnn._make_handle(context)
  File "/u/lamblinp/code/Theano/theano/gpuarray/dnn.py", line 85, in _make_handle
    raise RuntimeError("error creating cudnn handle")
RuntimeError: error creating cudnn handle
```

This works fine in the old back-end.
It does not crash either on the master branch, although it claims to use cuDNN on an unsupported device
```
lamblinp@eos21 % THEANO_FLAGS=device=cuda1 python -c 'import theano'
Mapped name None to device cuda1: GeForce GTX 580
PCI Bus ID: 0000:01:00.0
Using cuDNN version 5103 on context None
```

```
lamblinp@eos21 % THEANO_FLAGS=device=cuda1,dnn.enabled=False python -c 'import theano' 
Mapped name None to device cuda1: GeForce GTX 580
PCI Bus ID: 0000:01:00.0
```
(even on a machine that does not have a supported device):
```
lamblinp@bart1 % THEANO_FLAGS=device=cuda1 python -c 'import theano'                                                                          12:33
Mapped name None to device cuda1: GeForce GTX 580
PCI Bus ID: 0000:05:00.0
Using cuDNN version 5103 on context None
```

**Comment by nouiz** [comment]
Good catch, I have this now fixed:

On a computer where device 0 is a GTX750:

$ THEANO_FLAGS=device=cuda0 python -c "import theano"
/Tmp/lisa/os_v5/anaconda/lib/python2.7/site-packages/skcuda/cublas.py:273:
UserWarning: creating CUBLAS context to get version number
  warnings.warn('creating CUBLAS context to get version number')
Using cuDNN version 5103 on context None
Mapped name None to device cuda0: GeForce GTX 750 (0000:05:00.0)

On a computer where device0 is a k6000 and device 1 a 580 that don't
support cudnn:

$ THEANO_FLAGS=device=cuda0 python -c "import theano"
Using cuDNN version 5103 on context None
Mapped name None to device cuda0: Quadro K6000 (0000:02:00.0)

bastienf@eos21:~/repos/Theano$ THEANO_FLAGS=device=cuda1 python -c "import
theano"
Can not use cuDNN on context None: Device not supported
Mapped name None to device cuda1: GeForce GTX 580 (0000:01:00.0)

**Comment by nouiz** [comment]
Result while enabling multiple context at the same time:

$ THEANO_FLAGS="contexts=dev0->cuda0;dev1->cuda1" python -c "import theano"
Using cuDNN version 5103 on context dev0
Mapped name dev0 to device cuda0: Quadro K6000 (0000:02:00.0)
Can not use cuDNN on context dev1: Device not supported
Mapped name dev1 to device cuda1: GeForce GTX 580 (0000:01:00.0)

**Comment by nouiz** [comment]
rebased and added a commit to fix jenkins error.

**Pull request merged**